daemon/loadbalancer.go

```diff
@@ -182,47 +191,89 @@ func (h *deleteServiceID) Handle(params DeleteServiceIDParams) middleware.Respon
 	defer d.loadBalancer.BPFMapMU.Unlock()
 
 	svc, ok := d.loadBalancer.SVCMapID[types.ServiceID(params.ID)]
+
 	if !ok {
 		return NewDeleteServiceIDNotFound()
 	}
 
 	// FIXME: How to handle error?
-	DeleteL3n4AddrIDByUUID(uint32(params.ID))
+	err := DeleteL3n4AddrIDByUUID(uint32(params.ID))
+
+	if err != nil {
+		log.Warningf("error, DeleteL3n4AddrIDByUUID failed: %s", err)
+	}
 
 	if err := h.d.svcDelete(svc); err != nil {
+		log.Warningf("DELETE /service/{id}: error deleting svc %v: %s", svc, err)
 		return apierror.Error(DeleteServiceIDFailureCode, err)
 	}
 
 	return NewDeleteServiceIDOK()
 }
 
-// Deletes a service by the frontend address
-func (d *Daemon) svcDeleteByFrontend(frontend *types.L3n4Addr) error {
-	d.loadBalancer.BPFMapMU.Lock()
-	defer d.loadBalancer.BPFMapMU.Unlock()
-
+func (d *Daemon) svcDeleteByFrontendLocked(frontend *types.L3n4Addr) error {
 	svc, ok := d.loadBalancer.SVCMap[frontend.SHA256Sum()]
 	if !ok {
 		return fmt.Errorf("Service frontend not found %+v", frontend)
 	}
 	return d.svcDelete(&svc)
 }
 
+// Deletes a service by the frontend address
+func (d *Daemon) svcDeleteByFrontend(frontend *types.L3n4Addr) error {
+	d.loadBalancer.BPFMapMU.Lock()
+	defer d.loadBalancer.BPFMapMU.Unlock()
+
+	return d.svcDeleteByFrontendLocked(frontend)
+}
+
 func (d *Daemon) svcDelete(svc *types.LBSVC) error {
+	if err := d.svcDeleteBPF(svc); err != nil {
+		return err
+	}
+	d.loadBalancer.DeleteService(svc)
+	return nil
+}
+
+func (d *Daemon) svcDeleteBPF(svc *types.LBSVC) error {
+	log.Debugf("deleting service %d from BPF maps", svc.FE.ID)
 	var svcKey lbmap.ServiceKey
 	if !svc.FE.IsIPv6() {
-		svcKey = lbmap.NewService4Key(svc.FE.IP, svc.FE.Port, 0)
+		svcKey = lbmap.NewService4Key(svc.FE.IP, svc.FE.Port, svc.FE.Slave)
 	} else {
-		svcKey = lbmap.NewService6Key(svc.FE.IP, svc.FE.Port, 0)
+		svcKey = lbmap.NewService6Key(svc.FE.IP, svc.FE.Port, svc.FE.Slave)
 	}
 
 	svcKey.SetBackend(0)
 
-	if err := lbmap.DeleteService(svcKey); err != nil {
-		return err
+	// Get count of backends from master.
+	val, err := svcKey.Map().Lookup(svcKey.ToNetwork())
+	if err != nil {
+		return fmt.Errorf("key %s is not in lbmap", svcKey.ToNetwork())
 	}
 
-	d.loadBalancer.DeleteService(svc)
+	vval := val.(lbmap.ServiceValue)
+	numBackends := uint16(vval.GetCount())
+
+	// ServiceKeys are unique by their slave number, which corresponds to the number of backends. Delete each of these.
+	for i := numBackends; i > 0; i-- {
```

Comment:
Do you need to delete all map entries?

Reply:
If we're deleting the service, we should clean all of the entries out of the BPF map. Why would we not want to delete all of the entries?